### PR TITLE
Fixed uploading files in nested objects

### LINKED ISSUE #73: Upload in nested fields : Function DocumentReference.set() called with invalid data. Unsupported field value: a custom File object
Hi there,

Thank you for this awesome plugin !

I am stuck when I try to upload files in nested objects, I get this error : 
`Function DocumentReference.set() called with invalid data. Unsupported field value: a custom File object`

My structure is as follows : 

```
classes
  |_ title
  |_ sections
    |_ mySection
      |_ items[]
        |_ image       <---- Here is my file
          |_ url
```

It'd work fine if my structure was like : 
```
classes
  |_ title
  |_ items[]
    |_ image
      |_ url
```

After a little bit digging into it, I think the problem is that the `parseDataAndUpload()` function only looks for file fields in the root of the created/edited object, but I'm not sure...

Can someone point me to a solution please ?

### PR DESCRIPTION
**Pull request by simokhalil**
Fixes #73 

The problem was that uploading files deeply nested in JSON structure ended with exception.

Fixed by calling the `parseDataAndUpload()` function recursively until having a `rawFile` object.

### CODE REVIEW

**Comment by simokhalil** [comment]
I also made the field name where the "download url" is stored dynamic. This is done on `src` for now, by the `parseDataField()` function.

FileInput and ImageInput from react-admin transform the file object by adding `rawFile` and eventually `title`.

So if your markup is like : 
```
<FileInput source="file" label="Array Form Files">
  <FileField source="url" title="title" />
</FileInput>
```

The resulting object would be : 
```
{
  rawFile: {...},
  title: '...',
  url: '...',
}
```

So the simple solution is to look for the key that is different from `rawFile` and `title`

**Comment by dhstack** [comment]
I am having this issue too. Thank you @simokhalil for the fix! Can you update the conflicting files so @benwinding can review and merge?

**Comment by dhstack** [comment]
Can confirm this fix works for nested map objects with dot notation. For example...

```
<ImageInput source="checkIn.failed.lesson.image" label="Lesson Image" accept="image/*">
  <ImageField source="src" title="title" />
</ImageInput>
```

**Comment by simokhalil** [comment]
Thank you @dhstack 
Conflicts have been resolved

Will look at the save path problem asap

**Comment by simokhalil** [comment]
@dhstack @benwinding 
Here you are, it's now fixed (sorry for the latency)

Files in nested objects will be uploaded to nested folders reflecting the object tree.
If there is an array, the array index will be included in the path as a folder (example: `posts/{postId}/items/0/file`, `posts/{postId}/items/1/file`)

**Comment by dhstack** [comment]
@simokhalil Thanks for getting the array issue figured out. However, it still looks like the `Firebase Storage` bucket is not being saved to that matching path. For example, my object is stored the Firestore database as `/content/{docId}/checkIn/failed/lesson/image`, but in Storage it just puts it in `/content/{docId}/image`, as seen in the screenshot below - even after using the latest code you provided.

image

I would have expected it to be stored using the same object tree as the database (e.g. `/content/{docId}/checkIn/failed/lesson/image`)

Can you confirm the `Firebase Storage` path is using the same object tree as `Cloud Firestore`?

File: src/providers/database/FirebaseClient.ts [hunk 1]
```diff
@@ -308,7 +316,10 @@ export class FirebaseClient implements IFirebaseClient {
     if (!hasRawFile) {
       return;
     }
-    ref.src = await this.uploadAndGetLink(ref.rawFile, docPath, fieldPath);
+
+    const sourceFieldName = Object.keys(ref).find((key) => key !== 'rawFile' && key !== 'title');
+
+    ref[sourceFieldName] = await this.uploadAndGetLink(ref.rawFile, docPath, fieldPath);
     delete ref.rawFile;
   }
```
Comment (dhstack) [inline src/providers/database/FirebaseClient.ts]: I image the `uploadAndGetLink` function also needs to reflect the new object tree.

**Comment by simokhalil** [comment]
@dhstack 
Oh that's weird !
But ok, will take a look at it this weekend

**Comment by simokhalil** [comment]
@dhstack this should ok now
would you mind giving it a try please ?

**Comment by dhstack** [comment]
@simokhalil Yeah, this is particularly complicated. It doesn't look like it's working though. I'm expecting the path in `Firebase Storage` for my object to be:
`/content/ahZvmkXfBM4hAoCPwLVN_2_4/checkIn/succeeded/lesson`

But intead the path when uploaded is:
`/content/ahZvmkXfBM4hAoCPwLVN_2_4/succeeded/succeeded/succeeded/lesson/lesson/lesson`

So the sub paths work, but looks like they're being repeated or something.

For reference, my document object looks like this (irrelevant parts omitted):
```
{
  week: 2,
  day: 4,
  checkIn: {
    succeeded: {
      lesson: {
        image: {
          src: '',
          title: ''
        }
      }
    }
  }
}
```

**Comment by benwinding** [comment]
Hey guys,

Thanks for your efforts, I used some of your changes but also ended up refactoring the nested upload logic. 

Functionality of nested files should be available in version `3.1.16`

Have a look at the unit-testing too, to see if it works the intended way. It turns out it was easier to divide the task into parsing and calculating the upload objects, then uploading separately:

https://github.com/benwinding/react-admin-firebase/blob/8b6fc0d3f8d9871ac3fddd505194111cec44290c/tests/file-parser.spec.ts#L26-L45

Cheers,
Ben

**Comment by dhstack** [comment]
@benwinding Thank you for implementing this! Everything seems to work on my end.

**Pull request merged**